### PR TITLE
8339320: Optimize ClassFile Utf8EntryImpl#inflate

### PR DESCRIPTION
**Pull request by wenshao**
A very small optimization, split the large method inflate, split the infrequently used paths into a method inflateCHAR

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8339320](https://bugs.openjdk.org/browse/JDK-8339320): Optimize ClassFile Utf8EntryImpl#inflate (**Sub-task** - P4)

### Reviewers
 * [Chen Liang](https://openjdk.org/census#liach) (@liach - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20767/head:pull/20767` \
`$ git checkout pull/20767`

Update a local copy of the PR: \
`$ git checkout pull/20767` \
`$ git pull https://git.openjdk.org/jdk.git pull/20767/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20767`

View PR using the GUI difftool: \
`$ git pr show -t 20767`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20767.diff">https://git.openjdk.org/jdk/pull/20767.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20767#issuecomment-2321092999)

### CODE REVIEW

**Review by liach — CHANGES_REQUESTED** [review]
Java's UTF8 entry cannot use the 4-byte format in regular UTF8. You can check out how DataInputStream read UTF8 strings instead.

File: src/java.base/share/classes/jdk/internal/classfile/impl/AbstractPoolEntry.java [hunk 1]
```diff
@@ -382,29 +391,16 @@ else if ((state == State.STRING && u.state == State.STRING))
...
-                    return true;
+            if (charLen != s.length())
+                return false;
+            if (state == State.STRING)
+                return s.equals(stringValue);
+
+            if (state == State.BYTE) {
+                return JLA.regionMatches(s, rawBytes, LATIN1, offset, charLen);
+            } else {
+                return JLA.regionMatches(s, chars, UTF16, 0, charLen);
```
[the remark this answers is not in the document]
Comment (liach) [reply src/java.base/share/classes/jdk/internal/classfile/impl/AbstractPoolEntry.java:402]: Just looked at `ArraysSupport`; we don't need to ask the VM; we can already duplicate `mismatch(char[], int, char[], int, int)`, make one of its `char[]` parameter `byte[]` and proceed.

File: src/java.base/share/classes/jdk/internal/classfile/impl/AbstractPoolEntry.java [hunk 2]
```diff
@@ -231,67 +231,69 @@ private void inflate() {
                 this.hash = hashString(hash);
                 charLen = rawLen;
                 state = State.BYTE;
+            } else {
+                inflateHCAR(singleBytes, hash);
```
Comment (ExE-Boss) [inline src/java.base/share/classes/jdk/internal/classfile/impl/AbstractPoolEntry.java:235]:
```suggestion
                inflateCHAR(singleBytes, hash);
```

File: src/java.base/share/classes/jdk/internal/classfile/impl/AbstractPoolEntry.java [hunk 3]
```diff
@@ -231,67 +231,69 @@ private void inflate() {
...
-                    switch (c >> 4) {
-                        case 0, 1, 2, 3, 4, 5, 6, 7: {
-                            // 0xxx xxxx
-                            px++;
-                            chararr[chararr_count++] = (char) c;
-                            hash = 31 * hash + c;
-                            break;
+        }
+
+        private void inflateHCAR(int singleBytes, int hash) {
```
Comment (ExE-Boss) [inline src/java.base/share/classes/jdk/internal/classfile/impl/AbstractPoolEntry.java:239]:
```suggestion
        private void inflateCHAR(int singleBytes, int hash) {
```

**Review by ExE-Boss — COMMENTED** [review]
Typo (`HCAR` → `CHAR`):

File: src/java.base/share/classes/jdk/internal/classfile/impl/AbstractPoolEntry.java [hunk 4]
```diff
@@ -236,66 +236,69 @@ private void inflate() {
...
-                    switch (c >> 4) {
-                        case 0, 1, 2, 3, 4, 5, 6, 7: {
-                            // 0xxx xxxx
-                            px++;
-                            chararr[chararr_count++] = (char) c;
-                            hash = 31 * hash + c;
-                            break;
+        }
+
+        private void inflateCHAR(int singleBytes, int hash) {
```
Comment (liach) [inline src/java.base/share/classes/jdk/internal/classfile/impl/AbstractPoolEntry.java:244]: Can you rename this method to `inflateNonAscii`?

**Comment by liach** [comment]
Thanks for this rename and cleanup. Is `inflateNonAscii` code size less than 325?

**Comment by wenshao** [comment]
I don't have a test scenario for inflateNonAscii, and I can't see the codeSize of inflateNonAscii printed in the compiler log.

File: src/java.base/share/classes/jdk/internal/classfile/impl/AbstractPoolEntry.java [hunk 5]
```diff
@@ -249,66 +249,69 @@ private void inflate() {
...
-                    switch (c >> 4) {
-                        case 0, 1, 2, 3, 4, 5, 6, 7: {
-                            // 0xxx xxxx
-                            px++;
-                            chararr[chararr_count++] = (char) c;
-                            hash = 31 * hash + c;
-                            break;
+        }
+
+        private void inflateNonAscii(int singleBytes, int hash) throws ConstantPoolException {
```
Comment (liach) [inline src/java.base/share/classes/jdk/internal/classfile/impl/AbstractPoolEntry.java:257]:
```suggestion
        private void inflateNonAscii(int singleBytes, int hash) {
```

**Review by liach — APPROVED** [review]
We can optimize non-ascii reading here and in DataInputStream together later.

**Pull request closed without merging**